src/API/ProductsLowInStock.php

```diff
@@ -0,0 +1,309 @@
+		$response->header( 'X-WP-TotalPages', $query_results['pages'] );
+		$response->header( 'Cache-Control', 'max-age=300' );
+
+		return $response;
+	}
+
+	/**
+	 * Set the last order date for each data.
+	 *
+	 * @param array $results query result from get_low_in_stock_products.
+	 *
+	 * @return mixed
+	 */
+	protected function set_last_order_date( $results = array() ) {
+		global $wpdb;
+		if ( 0 === count( $results ) ) {
+			return $results;
+		}
+
+		$wheres = array();
+		foreach ( $results as $result ) {
+			'variation_product' === $result->post_type ?
+				array_push( $wheres, "(product_id={$result->post_parent} and variation_id={$result->ID})" )
+				: array_push( $wheres, "product_id={$result->ID}" );
+		}
+
+		count( $wheres ) ? $where_clause = implode( ' or ', $wheres ) : $where_clause = $wheres[0];
+
+		$product_lookup_table = $wpdb->prefix . 'wc_order_product_lookup';
+		$query_string         = "
+			select product_id, MAX( wc_order_product_lookup.date_created ) AS last_order_date 
+			from {$product_lookup_table} wc_order_product_lookup
+			where {$where_clause}
+			group by product_id
+			order by date_created desc
+		";
+
+		// phpcs:ignore -- ignore prepare() warning as we're not using any user input here.
+		$query_results = $wpdb->get_results( $query_string, OBJECT_K );
+		foreach ( $query_results as $key => $query_result ) {
+			if ( isset( $results[ $key ] ) ) {
+				$results[ $key ]->last_order_date = $query_result->last_order_date;
+			}
+		}
+
+		return $results;
+	}
+
+	/**
+	 * Get low in stock products data.
+	 *
+	 * @param int    $page current page.
+	 * @param int    $per_page items per page.
+	 * @param string $status post status.
+	 *
+	 * @return array
+	 */
+	protected function get_low_in_stock_products( $page = 1, $per_page = 1, $status = 'publish' ) {
+		global $wpdb;
+
+		$offset                       = ( $page - 1 ) * $per_page;
+		$low_stock_threshold          = absint( max( get_option( 'woocommerce_notify_low_stock_amount' ), 1 ) );
+		$use_sitewide_stock_threshold = $this->is_using_sitewide_stock_threshold_only();
+
+		if ( $use_sitewide_stock_threshold ) {
+			$query_string = $this->get_query( false );
+		} else {
+			$query_string = $this->get_query();
+		}
+
+		$query_results = $wpdb->get_results(
+			// phpcs:ignore -- not sure why phpcs complains about this line when prepare() is used here.
+			$wpdb->prepare( $query_string, $status, $low_stock_threshold, $offset, $per_page ),
+			OBJECT_K
+		);
+
+		$total_results = $wpdb->get_var( 'SELECT FOUND_ROWS()' );
+
+		return array(
+			'results' => $query_results,
+			'total'   => (int) $total_results,
+			'pages'   => (int) ceil( $total_results / (int) $per_page ),
+		);
+	}
+
+	/**
+	 * Check to see if store is using sitewide threshold only. Meaning that it does not have any custom
+	 * stock threshold for a product.
+	 *
+	 * @return bool
+	 */
+	public function is_using_sitewide_stock_threshold_only() {
```

Comment:
Nitpick: Could this be declared a `private` or `protected` method?

Further nitpick: you could switch all of the `protected` methods to `private` to help avoid backwards incompatible issues with inheritance. More info https://developer.woocommerce.com/2020/10/30/developer-advisory-inheritance-in-php-8/

Reply:
Good catch 👍 updated in https://github.com/woocommerce/woocommerce-admin/pull/7377/commits/426ecb0a5405298cee0c469e620e1337b2d7b3d8

I think the problem described in the article is not related. It talks about when someone extends this new class and uses a mismatching method signature. I don't expect anyone to extend this class, but they should check the error in case they decide to extend this class.

Reply:
> I don't expect anyone to extend this class, but they should check the error in case they decide to extend this class

The problem is if a third party has extended the class, we can’t then go and change the method signature without breaking things for them because we will create the mismatching signature. Hope that makes sense.

> Converting these methods to private prevent extending this class. If that's the case, we can simply make the class final

I think we should prevent extending the class through inheritance unless there is a use-case for it. Good idea about `final`, I'm happy with that 🙂

Reply:
@adrianduffell Ah! I see what you mean now. Thank you for the explanation. Yup, `final` sounds good. I'll make the change.

Reply:
updated in https://github.com/woocommerce/woocommerce-admin/pull/7377/commits/5865208985af1284606af63c221dfc5ec22d9143